buildSrc/src/main/java/com/google/firebase/gradle/plugins/GradleUtils.kt

```diff
@@ -56,3 +59,70 @@ fun Project.childFile(provider: Provider<File>, childPath: String) =
  * ```
  */
 fun File.listFilesOrEmpty() = listFiles().orEmpty()
+
+fun kotlinModuleName(project: Project): String {
+  val fullyQualifiedProjectPath = project.path.replace(":".toRegex(), "-")
+  return project.rootProject.name + fullyQualifiedProjectPath
+}
+
+fun setupStaticAnalysis(project: Project, library: FirebaseLibraryExtension) {
+  project.afterEvaluate {
+    configurations.all {
+      if ("lintChecks" == name) {
+        for (checkProject in library.staticAnalysis.androidLintCheckProjects) {
+          project.dependencies.add("lintChecks", project.project(checkProject!!))
+        }
+      }
+    }
+  }
+  project.tasks.register("firebaseLint") { dependsOn("lint") }
+  Coverage.apply(library)
+}
+
+fun getApiInfo(project: Project, srcDirs: Set<File>): TaskProvider<ApiInformationTask> {
+  val outputFile =
+    project.rootProject.file(
+      Paths.get(
+        project.rootProject.buildDir.path,
+        "apiinfo",
+        project.path.substring(1).replace(":", "_")
+      )
+    )
+  val outputApiFile = File(outputFile.absolutePath + "_api.txt")
+  val apiTxt =
+    if (project.file("api.txt").exists()) project.file("api.txt")
+    else project.file(project.rootDir.toString() + "/empty-api.txt")
+  val apiInfo =
+    project.tasks.register<ApiInformationTask>("apiInformation") {
```

Comment:
Besides `kotlinModuleName`, none of these should be in here. Having separate methods for them is perfect, but they should be private methods wherever they're being used. I see _why_ you did that- because it's a lot of duplicate work between `FirebaseJavaLibraryPlugin` and `FirebaseLibraryPlugin`. To fix this, we have a couple options:

1. Create separate Task files with preset conventions
 -> Not very intuitive, as these are the only places most of these tasks are used
2. Combine the plugins
 -> Not possible, as our needs are different from android plugins and java plugins
3. Provide a super class

Currently, `FirebaseLibraryPlugin` is more like `FirebaseAndroidPlugin`. We should provide either a parent class that applies common procedures, OR- we have a `FirebaseLibraryPlugin` that facilitates the application of either (`FirebaseAndroidPlugin` and `FirebaseJavaPlugin`) and handles the common configuration itself.

For clarification, `Utils` is meant for helper extensions and methods that could have use in multiple scenarios. While these methods have use in multiple places- it's still the same scenario.

FWIW, FirebaseLibrary deserves a redesign worthy of a design doc anyhow.

Reply:
I think a super class makes the most sense here, but would a rename to have `FirebaseLibraryPlugin` as the base class break some things?

Reply:
Inside buildsrc? Most definitely. Outside buildsrc? Not if we do it right- since we technically just expose a plug-in interface. Although with a redesign we may end up needing to break things anyhow ;)

Reply:
Well if changing names is breaking, for now I've refactored both plugins to use `BaseFirebaseLibraryPlugin` to get the methods out of Utils